### ISSUE TITLE
Oddities with MorphTo and related models.

### ISSUE THREAD

**Issue by gms8994** [nunomaduro/larastan#1866]
- Larastan Version: 2.9.1
- `--level` used: max
- Pull request with failing test:

### Description
Using the code directly below gives `model_b` as `mixed`. If I attempt to typehint the `model_b` typehint the way I _think_ it should be which would be `@return MorphTo<Model, ModelB>`, larastan/phpstan tells me 
```
Method App\Models\ModelA::model_b() 
should return Illuminate\Database\Eloquent\Relations\MorphTo<Illuminate\Database\Eloquent\Model, App\Models\ModelB>
  but returns Illuminate\Database\Eloquent\Relations\MorphTo<Illuminate\Database\Eloquent\Model, App\Models\ModelA>.
```

If I instead use `@return MorphTo<ModelB, ModelA>` then the types provided by `dumpType` in the controller provide the right values, but I still get 
```
Method App\Models\ModelA::model_b() 
should return Illuminate\Database\Eloquent\Relations\MorphTo<App\Models\ModelB, App\Models\ModelA> 
  but returns Illuminate\Database\Eloquent\Relations\MorphTo<Illuminate\Database\Eloquent\Model, App\Models\ModelA>.
```
in ModelA.

I'm not sure how to provide a PR given the things that I'm seeing are different based on how things are set up.

### Laravel code where the issue was found

```php
<?php

namespace App\Http\Controllers;

use App\Models\ModelA;

class TestController extends Controller
{
    public function __invoke() : void
    {
        ModelA::with('model_b')->get()->each(function ($model_a) {
            \PHPStan\dumpType($model_a); // `ModelA` when using the exact code provided 
            \PHPStan\dumpType($model_a->model_b); // `mixed` when using the exact code provided
        });
    }
}
```

```php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\MorphTo;

class ModelA extends Model
{
    public function model_b() : MorphTo
    {
        return $this->morphTo();
    }
}
```

```php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class ModelB extends Model
{
}
```

**Comment by calebdw** [comment]
In general, it's difficult to determine which model a `morphTo` should return because by definition it is dynamic from the database. If you're only returning one model from the `morphTo` then that probably shouldn't be a polymorphic relation.

I've found it's easier to define the generics on the other side of the relation (for `MorphOne` / `MorphMany`), although this doesn't work for many-to-many polymorphic types because again, it's dynamic.

Your first example is incorrect, the `TChildModel` is always the model you came from, so the generics would be:
```php
/** @return MorphTo<Model, ModelA> */
// or
/** @return MorphTo<Model, self> */
```

If you're trying to go through the `MorphTo` side, then you might be able to define a union type or a parent class / interface, but for most cases this likely wouldn't provided anymore specific types over Model and you'd still have to use `instanceof` checks anyway in the code if you're using behaviours to specific models.


For example, we have an Audit model that records any changes to our models (there's a lot of them), it would be impossible to determine which model is return statically. However, it is possible to go from User to Audits with the proper generics.

```php
class Audit extends Model
{
    public function auditable(): MorphTo
    {
        return $this->morphTo();
    }
}

class User extends Model
{
    /** @return MorphMany<Audit> */
    public function audits(): MorphMany
    {
        return $this->morphMany(Audit::class, 'auditable');
    }
}
```

**Comment by gms8994** [comment]
Appreciate the response! In this case, it's a library code that's using a `morphTo`, so I can't really do much about that part.

What I'm trying to fix specifically is the `mixed` return type by defining the `@return MorphTo<>` values. However, the examples you provided both still show `\PHPStan\dumpType($model_a->model_b);` as `mixed`, which makes it difficult to then provide any reasoning within other code as to what `model_b` actually is.

If it makes it any easier, I'm using owen-it/laravel-auditing and it's Audit trait on a custom Audit model (for reasons) and I'm implementing the `user` relation similarly to the library but trying to provide correct typehints so we can actually use information about the user that triggered the audit event.

**Comment by calebdw** [comment]
Hmmm, it looks like this is the current behaviour---we might could update this to return `Model|null`

https://github.com/larastan/larastan/blob/467113c58d110ad617cf9e07ff49b0948d1c03cc/tests/Type/data/model-properties-relations.php#L22

**Comment by canvural** [comment]
Hi,

You can use the [stub files](https://phpstan.org/user-guide/stub-files) to provide correct PHPDocs for 3rd party code.

We return `mixed` for morph to relations on purpose to avoid errors like `Access to undefined property ... on Model` etc.

**Comment by calebdw** [comment]
@canvural, 

> We return mixed for morph to relations on purpose to avoid errors like Access to undefined property ... on Model etc.

That makes more sense now, but I feel like you're just kicking the can further down the road (you'll still have to type this on higher PHPStan levels).

The more correct solution would be to return `Model|null` (which is what actually happens) and have users use the public model api (such as `->getKey()` instead of `->id`).

**Comment by canvural** [comment]
We might do that for v3

**Comment by gms8994** [comment]
I took a look at the MorphTo.stub file that was provided in the larastan repo to try and better understand what was happening. Looking at it, it's suggesting the related model would be my User model, and the child model would be the Audit (assuming I understand it at all correctly).

If I change my `user` relationship on the Audit model to `@return MorphTo<\App\Models\User, self>` then the rest of the code that is using the `$audit->user` relationship works correctly, including a `dumpType($audit->user)` correctly calling out that this is a User model.

However, the model itself says `Method App\Models\Audit::user() should return Illuminate\Database\Eloquent\Relations\MorphTo<App\Models\User, App\Models\Audit> but returns Illuminate\Database\Eloquent\Relations\MorphTo<Illuminate\Database\Eloquent\Model, App\Models\Audit>.` though I'm not sure why my `class User extends \Illuminate\Foundation\Auth\User` wouldn't qualify as a model.

Am I just trying to put phpstan/larastan in to a situation that it can't well handle right now? Because this seems like it should work, but something isn't fully connecting for me.

**Comment by gms8994** [comment]
@canvural @calebdw am I doing something unsupported here? I would think that this should work for this specific instance that I'm trying to work with.

**Comment by olivernybroe** [comment]
We are facing the same issue here with it returning `mixed` instead of `Model`. 

Would be nice if it just used the `TRelatedModel` template, and then let the user set it to `mixed` or `Collection` or whatever fits their scenario.

**Comment by canvural** [comment]
@olivernybroe This was changed a while ago I think. We don't make it return mixed anymore both in 2.x and 3.x. [This](https://github.com/larastan/larastan/blob/87042f7ff4d13fc3fbfa2250ed888c0dd3a85eba/tests/Type/data/model-properties-relations.php#L93) is the test we have which has `Model|null`

Can you share more info?

**Comment by olivernybroe** [comment]
@canvural Yeah sure, so I have a model with a `morhpto` relation on it

```php
/**
 * @return MorphTo<\Illuminate\Database\Eloquent\Model, $this>
 */
public function resource(): MorphTo
{
    return $this->morphTo();
}
```

And doing the following in my tests
```php
$provider = NotificationProvider::query()->firstOrFail();
PHPStan\dumpType($provider);
PHPStan\dumpType($provider->resource());
PHPStan\dumpType($provider->resource);
```

Gives me 
```
  36     Dumped type: MyPackage\Models\NotificationProvider                                                                                      
  37     Dumped type: Illuminate\Database\Eloquent\Relations\MorphTo<Illuminate\Database\Eloquent\Model, MyPackage\Models\NotificationProvider>  
  38     Dumped type: mixed 
 ```

This is on 
v11.34.2 and v3.0.2

**Comment by calebdw** [comment]
This is why it's returning `mixed`:

https://github.com/larastan/larastan/blob/ae431fad4c2f6f8a34e7e297e19b9575a728ad85/src/Properties/ModelRelationsExtension.php#L154-L156

I had fixed this in #1990, but it didn't make the cherry-picking cut I guess

**Comment by calebdw** [comment]
Fixed in #2138

**Comment by olivernybroe** [comment]
Awesome! Thank you so much for fixing it 💯
